libs/architectures/architectures/networks/s4.py

```diff
@@ -0,0 +1,244 @@
+        C = torch.view_as_complex(self.C)  # (H N)
+        A = -torch.exp(self.log_A_real) + 1j * self.A_imag  # (H N)
+
+        # Vandermonde multiplication
+        dtA = A * dt.unsqueeze(-1)  # (H N)
+        K = dtA.unsqueeze(-1) * torch.arange(L, device=A.device)  # (H N L)
+        C = C * (torch.exp(dtA) - 1.0) / A
+        K = 2 * torch.einsum("hn, hnl -> hl", C, torch.exp(K)).real
+
+        return K
+
+    def register(self, name, tensor, lr=None):
+        """
+        Register a tensor with a configurable learning rate
+        and 0 weight decay
+        """
+
+        if lr == 0.0:
+            self.register_buffer(name, tensor)
+        else:
+            self.register_parameter(name, nn.Parameter(tensor))
+
+            optim = {"weight_decay": 0.0}
+            if lr is not None:
+                optim["lr"] = lr
+            setattr(getattr(self, name), "_optim", optim)
+
+
+class S4D(nn.Module):
+    def __init__(
+        self, d_model, d_state=64, dropout=0.0, transposed=True, **kernel_args
+    ):
+        super().__init__()
+
+        self.h = d_model
+        self.n = d_state
+        self.d_output = self.h
+        self.transposed = transposed
+
+        self.D = nn.Parameter(torch.randn(self.h))
+
+        # SSM Kernel
+        self.kernel = S4DKernel(self.h, N=self.n, **kernel_args)
+
+        # Pointwise
+        self.activation = nn.GELU()
+        # dropout_fn = nn.Dropout2d  # NOTE: bugged in PyTorch 1.11
+        dropout_fn = DropoutNd
+        self.dropout = dropout_fn(dropout) if dropout > 0.0 else nn.Identity()
+
+        # position-wise output transform to mix features
+        self.output_linear = nn.Sequential(
+            nn.Conv1d(self.h, 2 * self.h, kernel_size=1),
+            nn.GLU(dim=-2),
+        )
+
+    def forward(
+        self, u, **kwargs
+    ):  # absorbs return_output and transformer src mask
+        """Input and output shape (B, H, L)"""
+        if not self.transposed:
+            u = u.transpose(-1, -2)
+        L = u.size(-1)
+
+        # Compute SSM Kernel
+        k = self.kernel(L=L)  # (H L)
+
+        # Convolution
+        k_f = torch.fft.rfft(k, n=2 * L)  # (H L)
+        u_f = torch.fft.rfft(u, n=2 * L)  # (B H L)
+        y = torch.fft.irfft(u_f * k_f, n=2 * L)[..., :L]  # (B H L)
+
+        # Compute D term in state space equation
+        # Essentially a skip connection
+        y = y + u * self.D.unsqueeze(-1)
+
+        y = self.dropout(self.activation(y))
+        y = self.output_linear(y)
+        if not self.transposed:
+            y = y.transpose(-1, -2)
+        # Return a dummy state to satisfy this repo's interface,
+        # but this can be modified
+        return (
+            y,
+            None,
+        )
+
+
+class S4Model(nn.Module):
+    def __init__(
+        self,
+        d_input,
+        d_output=10,
+        d_model=256,
+        d_state=64,
+        n_layers=4,
+        dropout=0.2,
+        prenorm=False,
+        **kernel_args,
```

Comment:
@wbenoit26 Was discussing with Christina, can we make the `lr`, `dt_min` and `dt_max` as explicit arguments here instead of hiding them behind `kernel_args`?

Reply:
Sure, I'll update this